// suf-regex/lib/utility.regex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function escapeRegExp(text) {
    return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
}
exports.escapeRegExp = escapeRegExp;
/**
 * Check whether text is a variable.
 */
function isVar(text) {
    return /^[\t ]*?\$[\w-]+:[\w\-%"']*/.test(text);
}
exports.isVar = isVar;
/**
 * Check whether text is a *
 */
function isStar(text) {
    return /^[\t ]*?\*/.test(text);
}
exports.isStar = isStar;
/**
 * Check whether text is class, id or placeholder
 */
function isClassOrId(text) {
    return /^[\t ]*[#\.%+]/.test(text);
}
exports.isClassOrId = isClassOrId;
/**
 * Check whether text is class, id or placeholder
 */
function isEmptyOrWhitespace(text) {
    return /^[\t ]*$/.test(text);
}
exports.isEmptyOrWhitespace = isEmptyOrWhitespace;
/**
 * Check whether text is a property
 */
function isProperty(text, empty) {
    if (empty) {
        return !/^[\t ]*[\w\-]+: *\S+/.test(text);
    }
    return /^[\t ]*[\w\-]+:/.test(text);
}
exports.isProperty = isProperty;
/**
 * Check whether text is a include
 */
function isInclude(text) {
    return /^[\t ]*@include/.test(text);
}
exports.isInclude = isInclude;
/**
 * Check whether text is a keyframe
 */
function isKeyframes(text) {
    return /^[\t ]*@keyframes/.test(text);
}
exports.isKeyframes = isKeyframes;
/**
 * Check whether text is a mixin
 */
function isMixin(text) {
    return /^[\t ]*=|^[\t ]*@mixin/.test(text);
}
exports.isMixin = isMixin;
/**
 * Check whether text is a each
 */
function isEach(text) {
    return /^[\t ]*@each/.test(text);
}
exports.isEach = isEach;
/**
 * Check whether text starts with &
 */
function isAnd(text) {
    return /^[\t ]*&/.test(text);
}
exports.isAnd = isAnd;
/**
 * Check whether text is at rule
 */
function isAtRule(text) {
    return /^[\t ]*@/.test(text);
}
exports.isAtRule = isAtRule;
/**
 * Check whether text is bracket selector
 */
function isBracketSelector(text) {
    return /^[\t ]*\[[\w=\-*"' ]*\]/.test(text);
}
exports.isBracketSelector = isBracketSelector;
/**
 * checks if text last char is a number
 * @param {String} text
 * @return {CompletionItem}
 */
function isNumber(text) {
    var reg = /[0-9]$/;
    return reg.test(text) && !text.includes('#');
}
exports.isNumber = isNumber;
/**
 * Check whether text starts with an html tag.
 */
function isHtmlTag(text) {
    var isTag = false;
    if (/^[\t ]*(a|abbr|address|area|article|aside|audio|b|base|bdi|bdo|blockquote|body|br|button|canvas|caption|cite|code|col|colgroup|data|datalist|dd|del|details|dfn|dialog|div|dl|dt|em|embed|fieldset|figure|footer|form|h1|h2|h3|h4|h5|h6|head|header|hgroup|hr|html|i|iframe|img|input|ins|kbd|keygen|label|legend|li|link|main|map|mark|menu|menuitem|meta|meter|nav|noscript|object|ol|optgroup|option|output|p|param|pre|progress|q|rb|rp|rt|rtc|ruby|s|samp|script|section|select|small|source|span|strong|style|sub|summary|sup|svg|table|tbody|td|template|textarea|tfoot|th|thead|time|title|tr|track|u|ul|var|video|wbr)((:|::|,|\.|#)[:$#{}()\w\-\[\]='",\.# ]*)?$/.test(text)) {
        isTag = true;
    }
    return isTag;
}
exports.isHtmlTag = isHtmlTag;
/**
 * Check whether text starts with a self closing html tag.
 */
function isVoidHtmlTag(text) {
    var isTag = false;
    if (/^[\t ]*(area|base|br|col|embed|hr|img|input|link|meta|param|source|track|wbr|command|keygen|menuitem)((:|::|,|\.|#)[:$#{}()\w\-\[\]='",\.# ]*)?$/.test(text)) {
        isTag = true;
    }
    return isTag;
}
exports.isVoidHtmlTag = isVoidHtmlTag;
/**
 * Check whether text starts with ::.
 */
function isPseudo(text) {
    return /^[\t ]*::?/.test(text);
}
exports.isPseudo = isPseudo;
/**
 * Check whether text starts with @if.
 */
function isIfOrElse(text) {
    return /^[\t ]*@if|^ *@else/.test(text);
}
exports.isIfOrElse = isIfOrElse;
/**
 * Check whether text starts with @else.
 */
function isElse(text) {
    return /^[\t ]*@else/.test(text);
}
exports.isElse = isElse;
/**
 * Check whether text starts with //R.
 */
function isReset(text) {
    return /^[\t ]*\/?\/\/ *R *$/.test(text);
}
exports.isReset = isReset;
/**
 * Check whether text starts with //I.
 */
function isIgnore(text) {
    return /^[\t ]*\/?\/\/ *I *$/.test(text);
}
exports.isIgnore = isIgnore;
/**
 * Check whether text starts with //S.
 */
function isSassSpace(text) {
    return /^[\t ]*\/?\/\/ *S *$/.test(text);
}
exports.isSassSpace = isSassSpace;
/**
 * TODO
 */
function isPath(text) {
    return /^.*['"]\.?[\.\/]$/.test(text);
}
exports.isPath = isPath;
/**
 * TODO
 */
function isScssOrCss(text, wasLastLineCss) {
    if (wasLastLineCss === void 0) { wasLastLineCss = false; }
    if (wasLastLineCss && text.endsWith(',') && isClassOrId(text)) {
        return true;
    }
    // comments get handled somewhere else.
    return /[;\{\}][\t ]*(\/\/.*)?$/.test(text);
}
exports.isScssOrCss = isScssOrCss;
/**
 * TODO
 */
function isCssPseudo(text) {
    return /^[\t ]*[&.#%].*:/.test(text);
}
exports.isCssPseudo = isCssPseudo;
/**
 * TODO
 */
function isCssOneLiner(text) {
    return /^[\t ]*[&.#%][\w-]*(?!#)\{.*[;\}]$/.test(text);
}
exports.isCssOneLiner = isCssOneLiner;
/**
 * TODO
 */
function isPseudoWithParenthesis(text) {
    return /^[\t ]*::?[\w\-]+\(.*\)/.test(text);
}
exports.isPseudoWithParenthesis = isPseudoWithParenthesis;
/**
 * TODO
 */
function isComment(text) {
    return /^[\t ]*\/\/|^ *\/\*/.test(text);
}
exports.isComment = isComment;
/**
 * TODO
 */
function isBlockCommentStart(text) {
    return /^[\t ]*(\/\*)/.test(text);
}
exports.isBlockCommentStart = isBlockCommentStart;
/**
 * TODO
 */
function isBlockCommentEnd(text) {
    return /[\t ]*\*\/|(?=^[a-zA-Z0-9#.%$@\\[=*+])/.test(text);
}
exports.isBlockCommentEnd = isBlockCommentEnd;
/**
 * TODO
 */
function isMoreThanOneClassOrId(text) {
    return /^[\t ]*[\.#%].* ?, *[\.#%].*/.test(text);
}
exports.isMoreThanOneClassOrId = isMoreThanOneClassOrId;
/**
 * TODO
 */
function hasColor(text) {
    return /^.*#[a-fA-F\d]{3,4}\b|^.*#[a-fA-F\d]{6}\b|^.*#[a-fA-F\d]{8}\b|rgba?\([\d. ]+,[\d. ]+,[\d. ]+(,[\d. ]+)?\)/.test(text);
}
exports.hasColor = hasColor;
/**
 * TODO
 */
function isBracketOrWhitespace(text) {
    return /^[\t ]*[}{]+[\t }{]*/.test(text);
}
exports.isBracketOrWhitespace = isBracketOrWhitespace;
/**
 * if the Property Value Space is none or more that one, this function returns false, else true;
 */
function hasPropertyValueSpace(text) {
    var split = text.split(':');
    return split[1] === undefined
        ? true
        : split[1][0] === undefined
            ? true
            : split[1].startsWith(' ')
                ? split[1][1] === undefined
                    ? true
                    : !split[1][1].startsWith(' ')
                : false;
}
exports.hasPropertyValueSpace = hasPropertyValueSpace;
/**
 * returns the distance between the beginning and the first char that is not the checkAgainstChar in form of a number.
 * @param checkAgainstChar defaults to `' '` should always be only one char.
 */
function getDistance(text, tabSize) {
    var count = 0;
    for (var i = 0; i < text.length; i++) {
        var char = text[i];
        if (char !== ' ' && char !== '\t') {
            break;
        }
        if (char === '\t') {
            count += tabSize;
        }
        else {
            count++;
        }
    }
    return count;
}
exports.getDistance = getDistance;
/**
 * returns the distance between the end and the first char that is not the checkAgainstChar in form of a number.
 * @param checkAgainstChar defaults to `' '` should always be only one char.
 */
function getDistanceReversed(text, tabSize) {
    var count = 0;
    for (var i = text.length - 1; i > 0; i--) {
        var char = text[i];
        if (char !== ' ' && char !== '\t') {
            break;
        }
        if (char === '\t') {
            count += tabSize;
        }
        else {
            count++;
        }
    }
    return count;
}
exports.getDistanceReversed = getDistanceReversed;
/**
 * TODO
 */
function splitOnce(text, splitter) {
    var split = text.split(splitter);
    var key = split.shift();
    return { body: (split.length > 0 ? splitter : '') + split.join(splitter), key: key };
}
exports.splitOnce = splitOnce;
